gae/room/common.py

```diff
@@ -88,76 +40,182 @@
-
-
-def GetUser(request):
-  """Determines the current user and loads some state.
+def get_all_programs_and_seed_data_if_necessary():
+  """
+  this method's raison d'etre is to create previously hard-coded programs and program types:
+    (a) in the prod datastore during the data migration of November of 2017, and
+    (b) in test datastores for test classes
 
-  In production, gets user from appengine.api.users.get_current_user(). So,
-  it should be used in App Engine handlers with login: required.
+  :return: tuple of program types created and programs created
+  :rtype: list[ndb_models.Program]
+  """
+  PROGRAMS = [
+    '2018 NRD',
+    '2018 Safe',
+    '2018 Teambuild',
+
+    '2017 NRD',
+    '2017 Safe',
+    '2017 Teambuild',
+
+    '2016 NRD',
+    '2016 Misc',
+    '2016 Safe',
+    '2016 Energy',
+    '2016 Teambuild',
+    '2016 Youth',
+
+    '2015 NRD',
+    '2015 Misc',
+    '2015 Safe',
+    '2015 Energy',
+    '2015 Teambuild',
+    '2015 Youth',
+
+    '2014 NRD',
+    '2014 Misc',
+    '2014 Safe',
+    '2014 Energy',
+    '2014 Teambuild',
+    '2014 Youth',
+
+    '2013 NRD',
+    '2013 Misc',
+    '2013 Safe',
+    '2013 Energy',
+    '2013 Teambuild',
+    '2013 Youth',
+
+    '2012 NRD',
+    '2012 Misc',
+    '2012 Safe',
+    '2012 Energy',
+    '2012 Teambuild',
+    '2012 Youth',
+
+    '2011 NRD',
+    '2011 Misc',
+    '2011 Safe',
+    '2011 Energy',
+    '2011 Teambuild',
+    '2011 Youth',
+
+    '2011 Test',
+
+    '2010 NRD',
+  ]
+
+  programs = ndb_models.Program.query().fetch()
+  if not programs:
+    # no programs exist yet, must seed
+    programs = []
+    assert not ndb_models.ProgramType.query().get()
+    program_data = dict()
+    for program_name in PROGRAMS:
+      year, program_type_name = program_name.split()
+      year = int(year)
+      status = ndb_models.Program.INACTIVE_STATUS
+      if year >= datetime.datetime.today().year:
+        status = ndb_models.Program.ACTIVE_STATUS
+      program_data[program_type_name] = {'year': year, 'status': status, 'program_type_key': None}
+
+    for program_type_name in program_data:
+      program_type, created = ndb_models.ProgramType.get_or_create(program_type_name)
+      assert created
+      program_data[program_type_name]['program_type_key'] = program_type.key
+
+    for program_type_name, program_datum in program_data.items():
+
+      program, created = ndb_models.Program.get_or_create(**program_datum)
+      assert created
+      programs.append(program)
+
+  return programs
 
-  Special handling for development server (dev_appserver.py):
+def IsDev():
+  return os.environ.get('SERVER_SOFTWARE', '').startswith('Development')
 
-  First, try to get user from the request, expecting an email address as
-  the value of the X-ROOMS_DEV_SIGNIN_EMAIL header.
 
-  Fall back to pre-configured identify from environment, meaning it was
-  probably set in app.yaml or unit test initialization like nose2-gae config.
+class RoomsUser(users.User):
```

Comment:
Not sure that inheritance is the right pattern here because it locks us into the AppEngine User model.  What I'm seeing in the code is that an AppEngine user is not sufficient for ROOMS and also a ROOMS user doesn't have to be an AppEngine user in some cases (test for example).  May be better to use a facade and pass through the few User attributes that are useful, like email.